### PR TITLE
Block global checkpoint advances when recovering

### PR DESCRIPTION
**Pull request by jasontedor**
After a replica shard finishes recovery, it will be marked as active and its local checkpoint will be considered in the calculation of the global checkpoint on the primary. If there were operations in flight during recovery, when the replica is activated its local checkpoint could be lagging behind the global checkpoint on the primary. This means that when the replica shard is activated, we can end up in a situtaion where a global checkpoint update would want to move the global checkpoint backwards, violating an invariant of the system. This only arises if a background global checkpoint sync executes, which today is only a scheduled operation and might be delayed until the in-flight operations complete and the replica catches up to the primary. Yet, we are going to move to inlining global checkpoints which will cause this issue to be more likely to manifest. Additionally, the global checkpoint on the replica, which is the local knowledge on the replica updated under the mandate of the primary, could be higher than the local checkpoint on the replica, again violating an invariant of the system. This commit addresses these issues by blocking global checkpoint on the primary when a replica shard is finalizing recovery. While we have blocked global checkpoint advancement, recovery on the replica shard will not be considered complete until its local checkpoint advances to the blocked global checkpoint.

Relates #10708

### CODE REVIEW

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left smaller comments and questions and will reach out to discuss them. Note that CI seems to have found an issue.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java [hunk 1]
```diff
@@ -173,33 +205,79 @@ synchronized void updateCheckpointOnReplica(final long checkpoint) {
...
+                }
+            }
+        } finally {
+            pendingInSync.remove(allocationId);
+        }
+    }
+
+    @SuppressForbidden(reason = "Object#wait for local checkpoint advancement")
+    private synchronized void waitForLocalCheckpointToAdvance() throws InterruptedException {
+        this.wait(TimeUnit.SECONDS.toMillis(30));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: That's an artifact of testing, I will remove.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -1516,11 +1517,17 @@ public void updateGlobalCheckpointOnPrimary() {
...
-    public void updateGlobalCheckpointOnReplica(final long checkpoint) {
+    public void updateGlobalCheckpointOnReplica(final long globalCheckpoint) {
         verifyReplicationTarget();
-        getEngine().seqNoService().updateGlobalCheckpointOnReplica(checkpoint);
+        final SequenceNumbersService seqNoService = getEngine().seqNoService();
+        final long localCheckpoint = seqNoService.getLocalCheckpoint();
+        if (globalCheckpoint <= localCheckpoint) {
+            seqNoService.updateGlobalCheckpointOnReplica(globalCheckpoint);
+        } else {
+            assert state() == IndexShardState.RECOVERING;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I agree, I will add a detailed comment and tighten the assertion.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java [hunk 3]
```diff
@@ -49,14 +48,20 @@
      * through recovery. These shards are treated as valid copies and participate in determining the global checkpoint. This map is keyed by
      * allocation IDs. All accesses to this set are guarded by a lock on this.
      */
-    private final ObjectLongMap<String> inSyncLocalCheckpoints;
+    final ObjectLongMap<String> inSyncLocalCheckpoints;
+
+    /*
+     * This map holds the last known local checkpoint for initializing shards that is undergoing recovery. Such shards do not participate
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I will fix this grammar issue.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java [hunk 4]
```diff
@@ -173,33 +205,79 @@ synchronized void updateCheckpointOnReplica(final long checkpoint) {
...
+
+        updateLocalCheckpoint(allocationId, localCheckpoint, trackingLocalCheckpoints, "tracking");
+        waitForAllocationIdToBeInSync(allocationId);
+    }
+
+    private synchronized void waitForAllocationIdToBeInSync(final String allocationId) throws InterruptedException {
+        pendingInSync.add(allocationId);
+        try {
+            while (true) {
+                final long current = trackingLocalCheckpoints.get(allocationId);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I don't think that we can make any assertions here, we would have to push too much knowledge down and I don't think it's worth it. I can leave a comment explaining the situation though.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java [hunk 5]
```diff
@@ -173,33 +205,79 @@ synchronized void updateCheckpointOnReplica(final long checkpoint) {
...
+                if (current >= globalCheckpoint) {
+                    logger.trace("marked [{}] as in-sync with local checkpoint [{}]", allocationId, current);
+                    trackingLocalCheckpoints.remove(allocationId);
+                    inSyncLocalCheckpoints.put(allocationId, current);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I pushed 6ac2706c140960964b06d4377ceeb203025fb256.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I pushed c028510a73836ed03b3198fd541c73560cb653f7.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I pushed b6a9ce6ff5c2a8b7d177faf1928f2ebd6cc23080.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I pushed 10c0f3c718c7e46c3dc7fde5e57222f60d4dcc32.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I pushed 162d8e71844f63b892be55422cad71b14a9f5bf0.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java [hunk 6]
```diff
@@ -173,33 +205,79 @@ synchronized void updateCheckpointOnReplica(final long checkpoint) {
...
+     * @param allocationId    the allocation ID of the shard to mark as in-sync
+     * @param localCheckpoint the current local checkpoint on the shard
+     *
+     * @throws InterruptedException if the thread is interrupted waiting for the local checkpoint on the shard to advance
      */
-    public synchronized void markAllocationIdAsInSync(final String allocationId) {
-        if (assignedAllocationIds.contains(allocationId) == false) {
-            // master has removed this allocation, ignore
+    public synchronized void markAllocationIdAsInSync(final String allocationId, final long localCheckpoint) throws InterruptedException {
+        if (trackingLocalCheckpoints.containsKey(allocationId) == false) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java]: I pushed 6ac2706c140960964b06d4377ceeb203025fb256.

**Comment by jasontedor** [comment]
> Note that CI seems to have found an issue.

I'm not sure what this is, there's insufficient logging to understand the issue. The logs shows this:

```
1> [2017-05-01T17:28:10,739][INFO ][o.e.i.r.IndexRecoveryIT  ] --> bump replica count
  1> [2017-05-01T17:28:10,739][INFO ][o.e.c.m.MetaDataUpdateSettingsService] [node_t0] updating number_of_replicas to [1] for indices [test-idx-1]
  1> [2017-05-01T17:28:10,958][INFO ][o.e.c.r.a.AllocationService] [node_t0] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[test-idx-1][0]] ...]).
```

at which point the test teardown executes from a tripped assertion:

```
   > Throwable #1: java.lang.AssertionError: 
   > Expected: <0>
   >      but: was <1>
   > 	at __randomizedtesting.SeedInfo.seed([1B2956BE26BD615E:B682E5EA13E5031C]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.indices.recovery.IndexRecoveryIT.testRerouteRecovery(IndexRecoveryIT.java:370)
   > 	at java.lang.Thread.run(Thread.java:745)
```

This is asserting that the recovery stats show there are no targets. Given that we got to green (so the recovery had to have finished), and we previously successfully asserted that the recovery state is done, it's not clear how this tripped with there being a recovery target still. I run this 256 times and then 512 times on my CI and it never reproduced. It's also unclear to me how this PR could have impacted this?

I wonder if we can proceed and if CI finds an issue increase the logging, and if it gets noisy silence the test?

**Comment by jasontedor** [comment]
@ywelsch I've responded to your feedback.

**Review by ywelsch — APPROVED** [review]
LGTM. On to the next one :-)

**Comment by jasontedor** [comment]
Thanks @ywelsch.

**Pull request merged**